client/hwlib/src/collectors/hardware_info.rs

```diff
@@ -0,0 +1,317 @@
+/* Copyright 2024 Canonical Ltd.
+ * All rights reserved.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Written by:
+ *        Nadzeya Hutsko <nadzeya.hutsko@canonical.com>
+ */
+
+use anyhow::Result;
+use smbioslib::{
+    SMBiosBaseboardInformation, SMBiosData, SMBiosEntryPoint32, SMBiosEntryPoint64,
+    SMBiosInformation, SMBiosProcessorInformation, SMBiosSystemChassisInformation,
+    SMBiosSystemInformation, SMBiosVersion,
+};
+use std::fs;
+use std::io::ErrorKind;
+use time::macros::format_description;
+use time::Date;
+
+use super::cpuid::CpuId;
+use super::cpuinfo::{CpuFrequency, CpuInfo};
+use crate::models::devices::{Bios, Board, Chassis, Processor};
+
+pub fn load_smbios_data(
+    entry_filepath: &'static str,
+    table_filepath: &'static str,
+) -> Option<SMBiosData> {
+    match table_load_from_device(entry_filepath, table_filepath) {
+        Ok(data) => Some(data),
+        Err(e) => {
+            eprintln!("failed to load SMBIOS data: {}.", e);
+            None
+        }
+    }
+}
+
+impl TryFrom<&SMBiosInformation<'_>> for Bios {
+    type Error = anyhow::Error;
+
+    fn try_from(bios_info: &SMBiosInformation) -> Result<Self> {
+        let release_date_str = bios_info.release_date().to_string();
+        let release_date_format = format_description!("[month]/[day]/[year]");
+        let release_date_parsed = Date::parse(&release_date_str, &release_date_format)?;
+        let output_format = format_description!("[year]-[month]-[day]");
+        let release_date = release_date_parsed.format(&output_format)?;
+        let firmware_revision = match (
+            bios_info.system_bios_major_release(),
+            bios_info.system_bios_minor_release(),
+        ) {
+            (Some(major), Some(minor)) => Some(format!("{}.{}", major, minor)),
+            _ => None,
+        };
+        Ok(Bios {
+            firmware_revision,
+            release_date: Some(release_date),
+            revision: Some(bios_info.version().to_string()),
+            vendor: bios_info.vendor().to_string(),
+            version: bios_info.version().to_string(),
+        })
+    }
+}
+
+/// Retrieve CPU information from SMBIOS
+impl TryFrom<(&SMBiosProcessorInformation<'_>, &'static str)> for Processor {
```

Comment:
I've updated the code to use `&Path` instead of `&' static str`. However, I'd like to keep the approach with `TryFrom` since we don't implement `new` method for other structs, but the `TryFrom` trait instead. I guess it matches what we're actually doing (trying to create `Processor` object using the provided data)

Reply:
Okay this sounds good to me (especially the use of the path-types :) )